// vdux-ui/lib/components/DropdownMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _document = require('vdux/document');

var _document2 = _interopRequireDefault(_document);

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

var _util = require('../util');

var _noop = require('@f/noop');

var _noop2 = _interopRequireDefault(_noop);

var _Menu = require('./Menu');

var _Menu2 = _interopRequireDefault(_Menu);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /**
                                                                                                                                                                                                                              * Imports
                                                                                                                                                                                                                              */

/**
 * Dropdown container component
 */

function render(_ref) {
  var props = _ref.props;
  var children = _ref.children;
  var _props$onDismiss = props.onDismiss;
  var onDismiss = _props$onDismiss === undefined ? _noop2.default : _props$onDismiss;
  var left = props.left;
  var top = props.top;
  var open = props.open;

  var restProps = _objectWithoutProperties(props, ['onDismiss', 'left', 'top', 'open']);

  return (0, _element2.default)(
    _Menu2.default,
    _extends({
      boxSizing: 'border-box',
      boxShadow: 'menu',
      rounded: true,
      py: 's',
      absolute: {
        left: left ? 0 : 'auto',
        right: left ? 'auto' : 0,
        top: top ? 'auto' : '100%',
        bottom: top ? '100%' : 'auto'
      },
      bgColor: 'white',
      color: 'text',
      column: true,
      hide: !open
    }, restProps, {
      'class': (0, _util.classes)(props.class, 'vui-dropdown-menu') }),
    children,
    open && (0, _element2.default)(_document2.default, { onClick: onDismiss, onKeypress: { esc: onDismiss } })
  );
}

/**
 * Exports
 */

exports.default = {
  render: render
};
