// vdux-ui/lib/components/ErrorTip.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

var _Tooltip = require('./Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /**
                                                                                                                                                                                                                              * Imports
                                                                                                                                                                                                                              */

/**
 * <ErrorTip/>: An error message styled tooltip
 */

function render(_ref) {
  var props = _ref.props;
  var message = props.message;

  var rest = _objectWithoutProperties(props, ['message']);

  return (0, _element2.default)(
    _Tooltip2.default,
    _extends({
      fs: 'xxs',
      p: '0px 20px',
      lh: '30px',
      bgColor: 'error',
      space: 10,
      maxWidth: 'none'
    }, rest),
    message
  );
}

/**
 * Exports
 */

exports.default = {
  render: render
};
