// vdux-ui/lib/components/MenuItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * Imports
                                                                                                                                                                                                                                                                   */

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

var _util = require('../util');

var _Block = require('./Block');

var _Block2 = _interopRequireDefault(_Block);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * getProps
 */

function getProps(props) {
  if (!props.color) props.color = 'text';
  if (!props.bgColor) props.bgColor = 'white';

  return props;
}

/**
 * MenuItem
 */

function render(_ref) {
  var props = _ref.props;
  var children = _ref.children;

  return (0, _element2.default)(
    _Block2.default,
    _extends({
      px: 'm',
      py: 's',
      pointer: true,
      m: 0,
      textDecoration: 'none'
    }, props, {
      'class': (0, _util.classes)(props.class, 'vui-menu-item') }),
    props.text || children
  );
}

/**
 * Exports
 */

exports.default = {
  getProps: getProps,
  render: render
};
