// vdux-ui/lib/components/ProgressBar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

var _Block = require('./Block');

var _Block2 = _interopRequireDefault(_Block);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /**
                                                                                                                                                                                                                              * Imports
                                                                                                                                                                                                                              */

/**
 * <ProgressBar/> compoonent
 */

function render(_ref) {
  var props = _ref.props;
  var _props$progress = props.progress;
  var progress = _props$progress === undefined ? 0 : _props$progress;
  var _props$barProps = props.barProps;
  var barProps = _props$barProps === undefined ? {} : _props$barProps;

  var rest = _objectWithoutProperties(props, ['progress', 'barProps']);

  return (0, _element2.default)(
    _Block2.default,
    _extends({ overflow: 'hidden', h: 20, mb: '1em', bgColor: 'rgba(black, 0.05)', relative: true }, rest),
    (0, _element2.default)(_Block2.default, _extends({
      tall: true,
      bgColor: 'primary',
      absolute: 'top 0px left 0px',
      transition: 'width .6s ease'
    }, barProps, {
      w: Math.round(progress) + '%' }))
  );
}

/**
 * Exports
 */

exports.default = {
  render: render
};
