// vdux-ui/lib/components/Spinner.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

var _util = require('../util');

var _Base = require('./Base');

var _Base2 = _interopRequireDefault(_Base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /**
                                                                                                                                                                                                                              * Imports
                                                                                                                                                                                                                              */

/**
 * getProps
 */

function getProps(props, context) {
  props.$spinnerAnimation = context.uiTheme.spinnerAnimation;
  return props;
}

/**
 * Spinner
 */

function render(_ref) {
  var props = _ref.props;
  var children = _ref.children;
  var $spinnerAnimation = props.$spinnerAnimation;
  var dark = props.dark;

  var rest = _objectWithoutProperties(props, ['$spinnerAnimation', 'dark']);

  return (0, _element2.default)(_Base2.default, _extends({
    tag: 'div',
    circle: '1em',
    bgColor: dark ? 'blue' : 'white',
    animation: dark ? $spinnerAnimation.dark : $spinnerAnimation.light
  }, rest, {
    'class': (0, _util.classes)(props.class, 'vui-spinner') }));
}

/**
 * Exports
 */

exports.default = {
  getProps: getProps,
  render: render
};
