// vdux-ui/lib/components/Toast.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

var _util = require('../util');

var _Block = require('./Block');

var _Block2 = _interopRequireDefault(_Block);

var _Icon = require('./Icon');

var _Icon2 = _interopRequireDefault(_Icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /**
                                                                                                                                                                                                                              * Imports
                                                                                                                                                                                                                              */

function onCreate(_ref) {
  var props = _ref.props;

  return props.$transition.didEnter();
}

/**
 * <Toast/> component
 */

function render(_ref2) {
  var props = _ref2.props;
  var children = _ref2.children;
  var _props$$transition = props.$transition;
  var $transition = _props$$transition === undefined ? {} : _props$$transition;
  var onDismiss = props.onDismiss;

  var rest = _objectWithoutProperties(props, ['$transition', 'onDismiss']);

  var entering = $transition.entering;
  var leaving = $transition.leaving;
  var didEnter = $transition.didEnter;
  var didLeave = $transition.didLeave;


  return (0, _element2.default)(
    _Block2.default,
    _extends({
      transform: 'translate3d(0, ' + (entering || leaving ? '-105% ' : '12px') + ', 0)',
      transition: 'transform .3s cubic-bezier(0.55, 0, 0.55, 0.2)',
      'class': (0, _util.classes)(props.class, 'vui-toast'),
      onTransitionEnd: leaving && didLeave,
      minHeight: '46px',
      bgColor: 'white',
      boxShadow: 'z3',
      z: 'overlay',
      fw: 'normal',
      w: '400px',
      mx: 'auto',
      rounded: true,
      fixed: true,
      right: true,
      left: true,
      top: true
    }, rest),
    (0, _element2.default)(_Icon2.default, {
      absolute: { top: -9, right: -9 },
      border: '2px solid white',
      align: 'center center',
      onClick: onDismiss,
      hide: !onDismiss,
      bgColor: 'black',
      boxShadow: 'z2',
      circle: '21px',
      color: 'white',
      name: 'close',
      pointer: true,
      fs: 'xs' }),
    (0, _element2.default)(
      _Block2.default,
      { m: '10px 12px', bold: true },
      children
    )
  );
}

/**
 * Exports
 */

exports.default = {
  onCreate: onCreate,
  render: render
};
