// vdux-ui/lib/components/Tooltip.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _vduxPosition = require('vdux-position');

var _vduxPosition2 = _interopRequireDefault(_vduxPosition);

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

var _util = require('../util');

var _Block = require('./Block');

var _Block2 = _interopRequireDefault(_Block);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /**
                                                                                                                                                                                                                              * Imports
                                                                                                                                                                                                                              */

/**
 * Constants
 */

var width = '6px';

/**
 * Tooltip
 */

function render(_ref) {
  var props = _ref.props;
  var children = _ref.children;
  var _props$placement = props.placement;
  var placement = _props$placement === undefined ? 'top' : _props$placement;
  var _props$space = props.space;
  var space = _props$space === undefined ? 0 : _props$space;
  var _props$color = props.color;
  var color = _props$color === undefined ? 'white' : _props$color;
  var show = props.show;
  var _props$bgColor = props.bgColor;
  var bgColor = _props$bgColor === undefined ? 'black' : _props$bgColor;

  var restProps = _objectWithoutProperties(props, ['placement', 'space', 'color', 'show', 'bgColor']);

  var margin = {};

  if (placement === 'top') margin.mt = '-3px';else if (placement === 'bottom') margin.mb = '-3px';else if (placement === 'right') margin.mr = '-3px';else if (placement === 'left') margin.ml = '-3px';

  return (0, _element2.default)(
    _vduxPosition2.default,
    { placement: placement, space: space, disable: !show },
    (0, _element2.default)(
      _Block2.default,
      _extends({
        absolute: true,
        userSelect: 'none'
      }, margin, {
        py: width,
        display: show ? 'block' : 'none',
        z: 'tooltip',
        'class': (0, _util.classes)(props.class, 'vui-tooltip') }),
      (0, _element2.default)(_Block2.default, _extends({
        absolute: true,
        sq: 0,
        borderColor: 'transparent',
        borderStyle: 'solid'
      }, getArrowStyle(placement, bgColor), {
        'class': 'vui-tooltip-arrow' })),
      (0, _element2.default)(
        _Block2.default,
        _extends({ 'class': 'vui-tooltip-inner', fs: 'xxs', py: 6, px: 9, maxWidth: 200, rounded: true, bgColor: bgColor, whiteSpace: 'nowrap', color: color }, restProps),
        children
      )
    )
  );
}

/**
 * Compute base styles
 */

function getArrowStyle(placement, color) {
  switch (placement) {
    case 'top':
      return {
        bottom: 0,
        left: '50%',
        marginLeft: '-' + width,
        borderWidth: width + ' ' + width + ' 0',
        borderTopColor: color,
        marginTop: '-3px'
      };
    case 'bottom':
      return {
        top: 0,
        left: '50%',
        marginLeft: '-' + width,
        borderWidth: '0 ' + width + ' ' + width,
        borderBottomColor: color
      };
    case 'right':
      return {
        top: '50%',
        right: '100%',
        marginTop: '-' + width,
        borderWidth: width + ' ' + width + ' ' + width + ' 0',
        borderRightColor: color
      };
    case 'left':
      return {
        top: '50%',
        left: '100%',
        marginTop: '-' + width,
        borderWidth: width + ' 0 ' + width + ' ' + width,
        borderLeftColor: color
      };
    default:
      throw new Error('Unknown tooltip placement: "' + placement + '"');
  }
}

/**
 * Exports
 */

exports.default = {
  render: render
};
