// delegant/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.delegateGlobal = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _proxyEvent = require('./proxy-event');

var _proxyEvent2 = _interopRequireDefault(_proxyEvent);

var _domEvents = require('@f/dom-events');

var _domEvents2 = _interopRequireDefault(_domEvents);

var _foreach = require('@f/foreach');

var _foreach2 = _interopRequireDefault(_foreach);

var _compose = require('@f/compose');

var _compose2 = _interopRequireDefault(_compose);

var _evStore = require('ev-store');

var _evStore2 = _interopRequireDefault(_evStore);

var _mapArray = require('@f/map-array');

var _mapArray2 = _interopRequireDefault(_mapArray);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } } /**
                                                                                                                                                                                                     * Imports
                                                                                                                                                                                                     */

/**
 * Delegator
 */

function delegant(rootNode) {
  var fn = arguments.length <= 1 || arguments[1] === undefined ? function (v) {
    return v;
  } : arguments[1];

  return _compose2.default.apply(undefined, _toConsumableArray((0, _mapArray2.default)(bind, _domEvents2.default)));

  function bind(name) {
    var handler = listener(name);
    rootNode.addEventListener(name, handler, true);
    return function () {
      return rootNode.removeEventListener(name, handler, true);
    };
  }

  function listener(name) {
    return function (e) {
      return bubble(name, e.target, e);
    };
  }

  function bubble(name, target, e) {
    var es = (0, _evStore2.default)(target);
    var handler = es[name];

    if (handler) {
      var _ret = function () {
        var event = new _proxyEvent2.default(e);
        event.currentTarget = target;

        'function' === typeof handler ? fn(handler(event)) : (0, _foreach2.default)(function (handler) {
          return fn(handler(event));
        }, handler);

        if (event._stopPropagation || event._stopImmediatePropagation) {
          return {
            v: undefined
          };
        }
      }();

      if ((typeof _ret === 'undefined' ? 'undefined' : _typeof(_ret)) === "object") return _ret.v;
    }

    if (target.parentNode && target !== rootNode && e.bubbles) {
      bubble(name, target.parentNode, e);
    }
  }
}

function delegateGlobal(node, fn) {
  var store = (0, _evStore2.default)(node);
  return _compose2.default.apply(undefined, _toConsumableArray((0, _mapArray2.default)(bind, _domEvents2.default)));

  function bind(name) {
    var handler = listener(name);
    node.addEventListener(name, handler, true);
    return function () {
      return node.removeEventListener(name, handler, true);
    };
  }

  function listener(name) {
    return function (e) {
      return (0, _foreach2.default)(function (handle) {
        return fn(handle(e));
      }, store[name]);
    };
  }
}

/**
 * Exports
 */

exports.default = delegant;
exports.delegateGlobal = delegateGlobal;
