// vdux-containers/lib/imageLoaded.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _handleActions;

var _handleActions2 = require('@f/handle-actions');

var _handleActions3 = _interopRequireDefault(_handleActions2);

var _createAction = require('@f/create-action');

var _createAction2 = _interopRequireDefault(_createAction);

var _loadImage = require('@f/load-image');

var _loadImage2 = _interopRequireDefault(_loadImage);

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var hasImage = typeof window !== 'undefined';

function imageLoaded(fn) {
  return function (Component) {
    return {
      initialState: function initialState() {
        return {
          loaded: false
        };
      },
      onCreate: _regenerator2.default.mark(function onCreate(_ref) {
        var props = _ref.props;
        var local = _ref.local;
        var url;
        return _regenerator2.default.wrap(function onCreate$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                url = fn(props);

                if (!(hasImage && url)) {
                  _context.next = 6;
                  break;
                }

                _context.next = 4;
                return (0, _loadImage2.default)(url);

              case 4:
                _context.next = 6;
                return local(loaded)();

              case 6:
              case 'end':
                return _context.stop();
            }
          }
        }, onCreate, this);
      }),
      render: function render(_ref2) {
        var props = _ref2.props;
        var state = _ref2.state;
        var children = _ref2.children;

        return (0, _element2.default)(
          Component,
          (0, _extends3.default)({}, props, { isLoaded: state.loaded }),
          children
        );
      },
      onUpdate: _regenerator2.default.mark(function onUpdate(prev, next) {
        var purl, nurl;
        return _regenerator2.default.wrap(function onUpdate$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                purl = fn(prev.props);
                nurl = fn(next.props);

                if (!(hasImage && nurl && purl !== nurl)) {
                  _context2.next = 9;
                  break;
                }

                _context2.next = 5;
                return next.local(unloaded)();

              case 5:
                _context2.next = 7;
                return (0, _loadImage2.default)(nurl);

              case 7:
                _context2.next = 9;
                return next.local(loaded)();

              case 9:
              case 'end':
                return _context2.stop();
            }
          }
        }, onUpdate, this);
      }),


      reducer: reducer
    };
  };
}

var loaded = (0, _createAction2.default)('Image loader: image loaded', null, function () {
  return { logLevel: 'debug' };
});
var unloaded = (0, _createAction2.default)('Image loader: image unloaded', null, function () {
  return { logLevel: 'debug' };
});

var reducer = (0, _handleActions3.default)((_handleActions = {}, (0, _defineProperty3.default)(_handleActions, unloaded, function (state) {
  return (0, _extends3.default)({}, state, {
    loaded: false
  });
}), (0, _defineProperty3.default)(_handleActions, loaded, function (state) {
  return (0, _extends3.default)({}, state, {
    loaded: true
  });
}), _handleActions));

exports.default = imageLoaded;
