// vdux-containers/lib/Input.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _handleActions2 = require('@f/handle-actions');

var _handleActions3 = _interopRequireDefault(_handleActions2);

var _createAction = require('@f/create-action');

var _createAction2 = _interopRequireDefault(_createAction);

var _CSSContainer = require('./CSSContainer');

var _CSSContainer2 = _interopRequireDefault(_CSSContainer);

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

var _vduxUi = require('vdux-ui');

var _wrap = require('./wrap');

var _wrap2 = _interopRequireDefault(_wrap);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function render(_ref) {
  var props = _ref.props;
  var state = _ref.state;
  var local = _ref.local;
  var children = _ref.children;
  var _props$ui = props.ui;
  var Ui = _props$ui === undefined ? _vduxUi.Input : _props$ui;
  var onInvalid = props.onInvalid;
  var onChange = props.onChange;
  var invalid = state.invalid;
  var message = state.message;


  return (0, _element2.default)(Ui, (0, _extends3.default)({
    invalid: invalid,
    message: message
  }, props, {
    onChange: [onChange, local(function (e) {
      return setValidity('');
    })],
    onInvalid: [onInvalid, local(function (e) {
      return setValidity(e.target.validationMessage);
    })] }));
}

var setValidity = (0, _createAction2.default)('<Input/>: setValidity');

var reducer = (0, _handleActions3.default)((0, _defineProperty3.default)({}, setValidity, function (state, message) {
  return {
    invalid: !!message,
    message: message
  };
}));

exports.default = (0, _wrap2.default)(_CSSContainer2.default)({
  reducer: reducer,
  render: render
});
