// vdux-containers/lib/subscribe.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends3 = require('babel-runtime/helpers/extends');

var _extends4 = _interopRequireDefault(_extends3);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _handleActions2 = require('@f/handle-actions');

var _handleActions3 = _interopRequireDefault(_handleActions2);

var _createAction = require('@f/create-action');

var _createAction2 = _interopRequireDefault(_createAction);

var _mapValues = require('@f/map-values');

var _mapValues2 = _interopRequireDefault(_mapValues);

var _element = require('vdux/element');

var _element2 = _interopRequireDefault(_element);

var _mapObj = require('@f/map-obj');

var _mapObj2 = _interopRequireDefault(_mapObj);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var update = (0, _createAction2.default)('vdux-subscribe: update', null, function () {
  return { logLevel: 'debug' };
});

function subscribe(subscriptions) {
  return function (Component) {
    return {
      initialState: function initialState(_ref) {
        var props = _ref.props;
        var local = _ref.local;
        var path = _ref.path;

        return (0, _mapObj2.default)(function (sub, key) {
          return sub.initialState ? sub.initialState(props, function (value) {
            return local(update)({ key: key, value: value });
          }, path) : {};
        }, subscriptions);
      },
      onCreate: _regenerator2.default.mark(function onCreate(_ref2) {
        var path = _ref2.path;
        var props = _ref2.props;
        var state = _ref2.state;
        var local = _ref2.local;
        return _regenerator2.default.wrap(function onCreate$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return (0, _mapValues2.default)(function (sub, key) {
                  return sub(props, state[key], function (value) {
                    return local(update)({ key: key, value: value });
                  }, path);
                }, subscriptions);

              case 2:
              case 'end':
                return _context.stop();
            }
          }
        }, onCreate, this);
      }),
      render: function render(_ref3) {
        var props = _ref3.props;
        var state = _ref3.state;
        var children = _ref3.children;

        return (0, _element2.default)(
          Component,
          (0, _extends4.default)({}, props, state),
          children
        );
      },
      onUpdate: _regenerator2.default.mark(function onUpdate(prev, _ref4) {
        var path = _ref4.path;
        var props = _ref4.props;
        var state = _ref4.state;
        var local = _ref4.local;
        return _regenerator2.default.wrap(function onUpdate$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return (0, _mapValues2.default)(function (sub, key) {
                  return sub(props, state[key], function (value) {
                    return local(update)({ key: key, value: value });
                  }, path);
                }, subscriptions);

              case 2:
              case 'end':
                return _context2.stop();
            }
          }
        }, onUpdate, this);
      }),


      reducer: reducer
    };
  };
}

var reducer = (0, _handleActions3.default)((0, _defineProperty3.default)({}, update, function (state, _ref5) {
  var key = _ref5.key;
  var value = _ref5.value;
  return (0, _extends4.default)({}, state, (0, _defineProperty3.default)({}, key, (0, _extends4.default)({}, state[key] || {}, value)));
}));

exports.default = subscribe;
