// vdux-ui/lib/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createEvent = exports.rgbaify = exports.classes = exports.mergeTheme = exports.getThemeProps = exports.highlight = exports.flexify = exports.setScaled = exports.setupStylePrefixer = exports.autoprefix = exports.borderSetter = exports.boolSetter = exports.positionSetter = exports.scaleSetter = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; }; /**
                                                                                                                                                                                                                                                   * Imports
                                                                                                                                                                                                                                                   */

var _normalizeBowser2 = require('@f/normalize-bowser');

var _normalizeBowser3 = _interopRequireDefault(_normalizeBowser2);

var _defaultTheme = require('./default-theme');

var _defaultTheme2 = _interopRequireDefault(_defaultTheme);

var _prefixStyle = require('@f/prefix-style');

var _prefixStyle2 = _interopRequireDefault(_prefixStyle);

var _extend = require('@f/extend');

var _extend2 = _interopRequireDefault(_extend);

var _bowser = require('bowser');

var _bowser2 = _interopRequireDefault(_bowser);

var _pick = require('@f/pick');

var _pick2 = _interopRequireDefault(_pick);

var _color = require('color');

var _color2 = _interopRequireDefault(_color);

var _has = require('@f/has');

var _has2 = _interopRequireDefault(_has);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Prefixer
 */

var prefixer = function prefixer(style) {
  return style;
};

function setupStylePrefixer(userAgent) {
  var _normalizeBowser = (0, _normalizeBowser3.default)(_bowser2.default._detect(userAgent));

  var browser = _normalizeBowser.browser;
  var version = _normalizeBowser.version;

  prefixer = (0, _prefixStyle2.default)(browser, version);
}

function autoprefix(styles) {
  return prefixer(styles);
}

/**
 * scaleSetter
 *
 * Set a style property from a given theme scale
 */

function scaleSetter(styleKey) {
  var themeScaleKey = arguments.length <= 1 || arguments[1] === undefined ? 'scale' : arguments[1];
  var defaultValue = arguments.length <= 2 || arguments[2] === undefined ? 'm' : arguments[2];

  if (Array.isArray(styleKey)) {
    return function (style, val, theme) {
      var scale = theme[themeScaleKey];

      if (val === true) {
        val = defaultValue;
      }

      for (var i = 0; i < styleKey.length; ++i) {
        setScaled(style, styleKey[i], val, theme[themeScaleKey]);
      }
    };
  } else {
    return function (style, val, theme) {
      return setScaled(style, styleKey, val === true ? defaultValue : val, theme[themeScaleKey]);
    };
  }
}

/**
 * boolSetter
 *
 * Set a constant value if the property is true,
 * do nothing otherwise.
 */

function boolSetter(styleKey, value) {
  return function (style, val) {
    return val && (style[styleKey] = value);
  };
}

/**
 * borderSetter
 *
 * Set a 1px solid border. If you pass a string, that string
 * is used as a color (indexed into your theme colors).
 */

function borderSetter(borderKey) {
  var colorKey = borderKey + 'Color';
  var styleKey = borderKey + 'Style';
  var widthKey = borderKey + 'Width';

  return function (style, val, _ref, _ref2) {
    var colors = _ref.colors;
    var _ref2$borderWidth = _ref2.borderWidth;
    var borderWidth = _ref2$borderWidth === undefined ? '1px' : _ref2$borderWidth;

    if (val) {
      if (typeof val === 'string') {
        // If the string being set has spaces in it,
        // that are not inside of a color string,
        // assume it's a shorthand for specifying all
        // the properties, rather than a named color
        if (/^[^\(]*\s/.test(val)) {
          var parts = val.split(' ');

          for (var i = 0; i < parts.length; i++) {
            var part = parts[i];
            if ((0, _has2.default)(part, colors)) {
              parts[i] = colors[part];
            }
          }

          style[borderKey] = parts.join(' ');
          return;
        } else {
          setScaled(style, colorKey, val, colors);
        }
      }

      style[styleKey] = 'solid';
      style[widthKey] = borderWidth;
    }
  };
}

/**
 *  position(obj, str)
 *
 * Set position on a style object
 *
 *  * obj - Object. Obj to set style on
 *  * props - Object with keys absolute|relative|fixed set to position strings of
 *            the form `bottom right` or `bottom 10px right 5px`.
 *            i.e. `(top|bottom) (n)? (left|right) (n)?`
 *  * scale - The scale from which to select sizes
 */

var posRe = /^(top|bottom)(?:\s(\-?\d+[a-zA-Z]+))?\s(left|right)(?:\s(\-?\d+[a-zA-Z]+))?$/;

function positionSetter(styleKey) {
  return function (style, val, _ref3) {
    var scale = _ref3.scale;

    style.position = styleKey;

    if (typeof val === 'string') {
      var parts = posRe.exec(val);

      setScaled(style, parts[1], parts[2] || 0, scale);
      setScaled(style, parts[3], parts[4] || 0, scale);
    } else if ((typeof val === 'undefined' ? 'undefined' : _typeof(val)) === 'object') {
      setScaled(style, 'top', val.top, scale);
      setScaled(style, 'left', val.left, scale);
      setScaled(style, 'right', val.right, scale);
      setScaled(style, 'bottom', val.bottom, scale);
    }
  };
}

/**
 * setScaled
 *
 * Set a value from a scale if the scale has
 * a corresponding key for that value
 */

function setScaled(obj, key, val, scale) {
  if (typeof val !== 'undefined') {
    obj[key] = scale && (0, _has2.default)(val, scale) ? scale[val] : val;
  }
}

/**
 * posString
 *
 * Generate a position string
 * given a position and a number
 */

function posString(pos, n) {
  switch (pos) {
    case 'top':
      return n + 'px 0 0 0';
    case 'right':
      return '0 ' + n + 'px 0 0';
    case 'bottom':
      return '0 0 ' + n + 'px 0';
    case 'left':
      return '0 0 0 ' + n + 'px';
  }
}

/**
 * highlight
 *
 * Takes in a color and if that color is dark, lightens it
 * and if it is light, darkens it. This allows you to make
 * nice rollover effects where an element appears to
 * highlight when the mouse is over it.
 */

function highlight(color) {
  var amount = arguments.length <= 1 || arguments[1] === undefined ? 0.1 : arguments[1];

  if (color === 'transparent') return color;

  var clr = (0, _color2.default)(color);

  return clr.light() ? clr.darken(amount).rgbaString() : clr.lighten(amount).rgbaString();
}

function getThemeProps(themeProps) {
  return function (props) {
    var context = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];
    var _context$uiTheme = context.uiTheme;
    var uiTheme = _context$uiTheme === undefined ? {} : _context$uiTheme;

    props.$theme = (0, _pick2.default)(themeProps, uiTheme, _defaultTheme2.default);
    return props;
  };
}

/**
 * mergeTheme(ctxTheme)
 *
 * Merge the contextual theme with the default theme. Memoize this
 * so that we allocate/extend every time we render any components,
 * especially since the contextual theme should change only extremely
 * rarely, if ever.
 */

var lastCtxTheme = null;
var lastMergedTheme = null;

function mergeTheme(ctxTheme) {
  if (lastCtxTheme === ctxTheme && lastMergedTheme) return lastMergedTheme;
  lastCtxTheme = ctxTheme;
  lastMergedTheme = (0, _extend2.default)({}, _defaultTheme2.default, ctxTheme);
  return lastMergedTheme;
}

/**
 * classes
 *
 * Takes two class arguments and if both are truthy,
 * returns an array of both, otherwise just returns
 * the truthy one
 */

function classes(a, b) {
  return a && b ? Array.isArray(a) ? a.concat(b) : [a, b] : a ? a : b;
}

/**
 * flexify
 *
 * Just adds 'flex-' to 'start' or 'end' so we can use
 * a nicer syntax with our elements
 */

function flexify(str) {
  return str === 'start' || str === 'end' ? 'flex-' + str : str;
}

var rgbaRe = /rgba\(([^,]+),([^,]+)\)/g;

function rgbaify(str, colors) {
  var parts = rgbaRe.exec(str);

  while (parts) {
    var _parts = parts;

    var _parts2 = _slicedToArray(_parts, 3);

    var replace = _parts2[0];
    var color = _parts2[1];
    var alpha = _parts2[2];


    if ((0, _has2.default)(color, colors)) {
      color = colors[color];
    }

    var newStr = (0, _color2.default)(color).alpha(Number(alpha)).rgbaString();
    str = str.replace(replace, newStr);
    parts = rgbaRe.exec(str);
  }

  return str;
}

/**
 * createEvent
 *
 * Cross-browser way to create events
 */

function createEvent(type, bubbles) {
  try {
    return new FocusEvent(type, { bubbles: bubbles });
  } catch (err) {
    var e = document.createEvent('CustomEvent');
    e.initCustomEvent(type, bubbles, true, {});
    return e;
  }
}

/**
 * Exports
 */

exports.scaleSetter = scaleSetter;
exports.positionSetter = positionSetter;
exports.boolSetter = boolSetter;
exports.borderSetter = borderSetter;
exports.autoprefix = autoprefix;
exports.setupStylePrefixer = setupStylePrefixer;
exports.setScaled = setScaled;
exports.flexify = flexify;
exports.highlight = highlight;
exports.getThemeProps = getThemeProps;
exports.mergeTheme = mergeTheme;
exports.classes = classes;
exports.rgbaify = rgbaify;
exports.createEvent = createEvent;
