// vdux-fire/lib/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.once = exports.firebaseSet = exports.update = exports.invalidate = exports.unsubscribe = exports.subscribe = undefined;

var _createAction = require('@f/create-action');

var _createAction2 = _interopRequireDefault(_createAction);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var subscribe = (0, _createAction2.default)('vdux-fireSummon: subscribe');
var unsubscribe = (0, _createAction2.default)('vdux-fireSummon: unsubscribe');
var invalidate = (0, _createAction2.default)('vdux-fireSummon: invalidate');
var update = (0, _createAction2.default)('vdux-fireSummon: update');
var firebaseSet = (0, _createAction2.default)('vdux-fireSummon: set');
var once = (0, _createAction2.default)('vdux-fireSummon: once');

exports.subscribe = subscribe;
exports.unsubscribe = unsubscribe;
exports.invalidate = invalidate;
exports.update = update;
exports.firebaseSet = firebaseSet;
exports.once = once;
